### PR TITLE
cmd,internal: move internal plugins to separate package

### PR DESCRIPTION
**Pull request by AlexNPavel**
**Description of the change:** This PR removes the "Basic Tests" and "OLM Integration Tests" scorecard tests from the main SDK binary and splits them off into `osdk-scorecard-basic` and `osdk-scorecard-olm`. This includes only the code and CI changes. There will be separate PRs to add the plugin builds to the Makefile and to add documentation on how to use the plugins.

**Motivation for the change:** Remove built-in tests from scorecard to migrate it to a fully plugin based design.

### CODE REVIEW

File: internal/pkg/scorecard/test_definitions.go [hunk 1]
```diff
@@ -123,29 +123,32 @@ func MergeSuites(suites []TestSuite) ([]TestSuite, error) {
 	mergedSuites := []TestSuite{}
 	for _, suiteSlice := range suiteMap {
 		testMap := make(map[string][]TestResult)
+		logs := ""
```
Comment (joelanford) [inline internal/pkg/scorecard/test_definitions.go]: This is probably a good opportunity to use [`strings.Builder`](https://golang.org/pkg/strings/#Builder)

File: internal/pkg/scorecard/test_definitions.go [hunk 2]
```diff
@@ -123,29 +123,32 @@ func MergeSuites(suites []TestSuite) ([]TestSuite, error) {
...
 		for _, suite := range suiteSlice {
 			for _, result := range suite.TestResults {
 				testMap[result.Test.GetName()] = append(testMap[result.Test.GetName()], result)
 			}
+			logs = logs + suite.Log
```
Comment (joelanford) [inline internal/pkg/scorecard/test_definitions.go]: Would it make sense to add something to separate each suite's logs? 

Maybe something like `logs = logs + fmt.Sprintf("--- %s logs ---\n%s\n\n", suite.Name, suite.Log)`

**Comment by joelanford** [comment]
@AlexNPavel With this change, will running `operator-sdk scorecard` in a new project do anything? Maybe if no plugins exist, we could log a message that tells users where they could find the basic and olm plugins?

We probably need a step-by-step document describing what an existing scorecard user needs to do to get back to the same set of tests that were running before this change. @hasbro17 maybe this would be another thing to include the migration doc updates?

File: internal/pkg/scorecard/plugin_runner.go [hunk 3]
```diff
@@ -0,0 +1,380 @@
...
+	proxyPodGlobal *v1.Pod
+	cleanupFns     []cleanupFn
+)
+
+const (
+	scorecardPodName       = "operator-scorecard-test"
+	scorecardContainerName = "scorecard-proxy"
+)
+
+func SetupAndRunPlugin() (*scapiv1alpha1.ScorecardOutput, error) {
```
[the remark this answers is not in the document]
Comment (AlexNPavel) [reply internal/pkg/scorecard/plugin_runner.go]: I don't think there were any changes in the copy paste other than changing the logging so that the suite's logs get directed to the SuiteResult's log field and other logging goes to stderr (and thus the main scorecard log).

I think it would be useful to decouple the function from viper. I'm not sure if it would really make much sense to separate the basic and olm suites into separate functions though. They have a pretty much identical setup and teardown and the only difference are ~20 lines at line 287. Other than that, the setup that's done for the tests is the same, as is the teardown.

File: internal/pkg/scorecard/test_definitions.go (hunk 2 above)
Comment (AlexNPavel) [reply internal/pkg/scorecard/test_definitions.go]: Yeah, that would be clearer

Comment (AlexNPavel) [reply internal/pkg/scorecard/test_definitions.go]: Actually, I'm not quite sure. What this function is merging is suites with the same name. The log field is in the SuiteResult field, so we already know which suite the log belongs to. This line is just aggregating the logs of multiple runs of the same suite.  Maybe just separating the logs for each run with a `\n---\n` could help with clarity.

File: internal/pkg/scorecard/plugin_runner.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (joelanford) [reply internal/pkg/scorecard/plugin_runner.go]: > They have a pretty much identical setup and teardown and the only difference are ~20 lines at line 287. Other than that, the setup that's done for the tests is the same, as is the teardown.

Could we extract that common setup and teardown code into separate helper functions, such that the Basic and OLM functions end up being pretty simple?

```go
type CommonOpts {
    CommonOpt1 bool
}

type BasicOpts struct {
    CommonOpts

    BasicOpt1 string
}

type OLMOpts struct {
    CommonOpts

    OLMOpt1 string
}

func Basic(opts BasicOpts) (*scapiv1alpha1.ScorecardOutput, error) {
    if err := setup(opts.CommonOpts); err != nil {
        return nil, err
    }
    output, err := doBasicTests(opts)
    if err != nil {
        return nil, err
    }
    if err := teardown(opts.CommonOpts); err != nil {
        return nil, err
    }
    return output, nil
}

func OLM(opts OLMOpts) (*scapiv1alpha1.ScorecardOutput, error) {
    if err := setup(opts.CommonOpts); err != nil {
        return nil, err
    }
    output, err := doOLMTests(opts)
    if err != nil {
        return nil, err
    }
    if err := teardown(opts.CommonOpts); err != nil {
        return nil, err
    }
    return output, nil
}

**Comment by AlexNPavel** [comment]
@joelanford I've updated the error message for the scorecard to have it print a link to the scorecard docs (which will be updated in a future PR to include instructions on how to run basic and olm plugins) if it fails to run the plugins

File: internal/pkg/scorecard/plugin_runner.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (AlexNPavel) [reply internal/pkg/scorecard/plugin_runner.go]: Yeah, that would be good to do in another PR. The current function is quite massive right now and should be modularized.

File: internal/pkg/scorecard/scorecard.go [hunk 4]
```diff
@@ -336,6 +60,9 @@ func runTests() ([]scapiv1alpha1.ScorecardOutput, error) {
 	if err != nil {
 		return nil, fmt.Errorf("failed to list files in %s/bin: %v", pluginDir, err)
 	}
+	if len(files) == 0 {
+		return nil, fmt.Errorf("No scorecard tests found in %s/bin", pluginDir)
```
Comment (joelanford) [inline internal/pkg/scorecard/scorecard.go]: With a just-scaffolded project and a user that has not read the docs and configured a plugin, this will be the error they get, right? Maybe add the docs link here as well?

Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: They'd probably get the "Plugin directory not found" error that occurs higher up in the function. Either way, any error that occurs in this function gets handled by https://github.com/operator-framework/operator-sdk/pull/1472/files#diff-972e6cd1e836a6eed2c842356fd9ac78R112, which prints a link to the scorecard documentation.

Comment (joelanford) [reply internal/pkg/scorecard/scorecard.go]: Ahh yeah, duh. Even better. Thanks for explaining.

**Comment by joelanford** [comment]
@estroz That's a great point. At the very least, this would make it more difficult for users to keep up with changes/improvements we make to the base plugins. It also makes the initial setup more difficult. 

A lot of the modularization work has already happened with the plugin architecture and the updates to make the basic and OLM plugins output their results in the common plugin output format.

So maybe we could to come up with a `Plugin` interface abstraction that allows us to generically add, filter, and run a set of plugins, where the underlying plugin implementations can support both built-in and file-based executable plugins?

**Comment by estroz** [comment]
@joelanford yes an interface like that sounds like the correct approach. Something like:

```Go
type Plugin interface {
    Name() string
    Run() (*scapiv1alpha1.ScorecardOutput, error)
}
```

Then the scorecard could have builtins implemented as `Plugin`s, and wrap each binary in `{plugin-dir}/bin` in a generic `Plugin` implementation that expects scorecard JSON output from the binary, as we have it now.

@AlexNPavel WDYT?

**Comment by AlexNPavel** [comment]
@estroz That sounds good. One change that we will be keeping from here (other than the various minor logging and design improvements discovered by working with actual binary plugins) is removing the built-in test flags from the main scorecard command. IMO those can be confusing to users due to the fact that they don't affect other plugins. Instead, we will just use viper and a config file for the built-in test's options. We will provide users with a config flag for the built-in tests as well as the `kubeconfig` flag, which users can use to set the `KUBECONFIG` env var for other plugins as well.

**Comment by AlexNPavel** [comment]
/hold

This PR will heavily redesigned

**Comment by AlexNPavel** [comment]
This PR and it's purpose have been changed and it is now ready for review again.

This PR moves the internal plugins into a separate package at `internal/pkg/scorecard/plugins` and moves helpers/structs shared by the internal plugins and the scorecard runner into `internal/pkg/scorecard/helpers`. This means that the internal plugins are not tied to the state of the plugin runner in `scorecard.go` and could be run independently. This also creates a new `interface.go` file to define interfaces to simplify running both internal and external packages. There shouldn't be too much of a difference from a user's perspective other than better handling of logging for both internal and external plugins.

/cc @joelanford @estroz @hasbro17

**Comment by AlexNPavel** [comment]
IMPORTANT: The files view of this PR hides the `scorecard.go` file because it is a large diff. It does have a lot of changes though, so please make sure to expand the diff for that file and review it.

File: internal/pkg/scorecard/interfaces.go [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+
+	scplugins "github.com/operator-framework/operator-sdk/internal/pkg/scorecard/plugins"
```
Comment (estroz) [inline internal/pkg/scorecard/interfaces.go]: Nit: re-organize SDK deps

File: internal/pkg/scorecard/interfaces.go [hunk 6]
```diff
@@ -0,0 +1,143 @@
...
+}
+
+func (p genericPlugin) Name() string {
+	return filepath.Base(p.binPath)
+}
+
+func (p genericPlugin) Run() scapiv1alpha1.ScorecardOutput {
+	// This should never error since we chdir in getPlugins()
+	if err := os.Chdir(viper.GetString(PluginDirOpt)); err != nil {
+		name := fmt.Sprintf("Failed Plugin: %s", filepath.Base(p.binPath))
```
Comment (estroz) [inline internal/pkg/scorecard/interfaces.go]: Abstract failed/invalid plugin code into functions as its more or less copied in several places.

File: internal/pkg/scorecard/interfaces.go [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+		return failedPlugin(name, description, logs)
+	}
+	stderrString := string(stderr.Bytes())
+	if len(stderrString) != 0 {
+		log.Warn(stderrString)
+	}
+	return result
+}
+
+var BasicTestsPlugin = internalPlugin{
```
Comment (estroz) [inline internal/pkg/scorecard/interfaces.go]: Do these need to be global?

File: internal/pkg/scorecard/plugins/plugin_runner.go [hunk 8]
```diff
@@ -0,0 +1,362 @@
...
+)
+
+const (
+	scorecardPodName       = "operator-scorecard-test"
+	scorecardContainerName = "scorecard-proxy"
+)
+
+var log *logrus.Logger
+
+func RunInternalPlugin(plugin PluginType, pluginViper *viper.Viper, logFile io.ReadWriter) (scapiv1alpha1.ScorecardOutput, error) {
```
Comment (estroz) [inline internal/pkg/scorecard/plugins/plugin_runner.go]: Nit: `pluginViper` -> `config`

File: internal/pkg/scorecard/scorecard.go [hunk 9]
```diff
@@ -475,7 +189,7 @@ func initConfig() error {
 
 func configureLogger() error {
 	if viper.GetString(OutputFormatOpt) == HumanReadableOutputFormat {
-		logReadWriter = os.Stdout
+		logReadWriter = os.Stderr
```
Comment (estroz) [inline internal/pkg/scorecard/scorecard.go]: What's the reason for this change?

File: internal/pkg/scorecard/scorecard.go [hunk 10]
```diff
@@ -99,277 +49,36 @@ var (
 	log           = logrus.New()
 )
 
-func runTests() ([]scapiv1alpha1.ScorecardOutput, error) {
-	defer func() {
-		if err := cleanupScorecard(); err != nil {
-			log.Errorf("Failed to cleanup resources: (%v)", err)
-		}
-	}()
+var rootDir string
```
Comment (estroz) [inline internal/pkg/scorecard/scorecard.go]: Any reason for this global? Why not call `os.Getwd()` where needed?

File: internal/pkg/scorecard/plugins/plugin_runner.go [hunk 11]
```diff
@@ -0,0 +1,362 @@
...
+	"k8s.io/apimachinery/pkg/runtime/serializer"
+	"k8s.io/client-go/discovery/cached"
+	"k8s.io/client-go/kubernetes"
+	cgoscheme "k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/restmapper"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
```
Comment (estroz) [inline internal/pkg/scorecard/plugins/plugin_runner.go]: It would be cleaner to move `*Opt` and `validateScorecardPluginFlags()` to a `internal/pkg/scorecard/plugins/config.go` file, to reduce clutter.

File: internal/pkg/scorecard/interfaces.go (hunk 6 above)
Comment (AlexNPavel) [reply internal/pkg/scorecard/interfaces.go]: I don't think this can really be abstracted more? There are slight differences between each of the 5 error cases to make the errors clearer to the user.

File: internal/pkg/scorecard/scorecard.go (hunk 10 above)
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: I use this variable in the `Run` function for `internalPlugin`. Since the external plugins run in a different directory, I automatically `chdir` to the `rootDir` before running internal plugins to make sure I'm in the right place.

File: internal/pkg/scorecard/interfaces.go (hunk 7 above)
Comment (AlexNPavel) [reply internal/pkg/scorecard/interfaces.go]: Nope. Fixed

File: internal/pkg/scorecard/scorecard.go (hunk 9 above)
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: This was a leftover change from when I was splitting the plugins from the sdk binary. This should be Stdout.

Fixed

File: internal/pkg/scorecard/scorecard.go [hunk 12]
```diff
@@ -394,15 +89,20 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
 		return err
 	}
 	cmd.SilenceUsage = true
-	pluginOutputs, err := runTests()
+	var err error
+	rootDir, err = os.Getwd()
 	if err != nil {
-		return err
+		return fmt.Errorf("faild to get current working directory: %v", err)
```
Comment (LiliC) [inline internal/pkg/scorecard/scorecard.go]:
```suggestion
		return fmt.Errorf("failed to get current working directory: %v", err)
```

File: internal/pkg/scorecard/scorecard.go [hunk 13]
```diff
@@ -394,15 +89,20 @@ func ScorecardTests(cmd *cobra.Command, args []string) error {
 		return err
 	}
 	cmd.SilenceUsage = true
-	pluginOutputs, err := runTests()
+	var err error
```
Comment (LiliC) [inline internal/pkg/scorecard/scorecard.go]: Why the error declaration before?

File: internal/pkg/scorecard/plugins/config.go [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+		return errors.New("at least one test type must be set")
+	}
+	if config.GetBool(OLMTestsOpt) && config.GetString(CSVPathOpt) == "" {
+		return fmt.Errorf("csv-path must be set if olm-tests is enabled")
+	}
+	if config.GetBool(OlmDeployedOpt) && config.GetString(CSVPathOpt) == "" {
+		return fmt.Errorf("csv-path must be set if olm-deployed is enabled")
+	}
+	pullPolicy := config.GetString(ProxyPullPolicyOpt)
+	if pullPolicy != "Always" && pullPolicy != "Never" && pullPolicy != "PullIfNotPresent" {
```
Comment (LiliC) [inline internal/pkg/scorecard/plugins/config.go]: Nit: We could use the constants here https://github.com/kubernetes/kubernetes/blob/016e9d5c06089774c6286fd825302cbae661a446//pkg/apis/core/types.go#L1889:2

File: internal/pkg/scorecard/scorecard.go (hunk 13 above)
Comment (AlexNPavel) [reply internal/pkg/scorecard/scorecard.go]: On the next line, we are setting a global variable, so we have to use `=`, which means that `err` needs to be declared. If we use `:=`, that creates a new local variable and doesn't change the global one.

Comment (LiliC) [reply internal/pkg/scorecard/scorecard.go]: Interesting! Could we add a comment? So someone doesn't remove that in the future as it's not super obvious, at least not to me.
